File: sdks/python/mallard_sdk/bucketing.py
"""Deterministic bucketing — byte-for-byte identical to the server.

See ``backend/app/assignment/bucketing.py`` and ``sdks/fixtures/bucketing.json``.
"""

from __future__ import annotations

import hashlib

_UINT64 = 2**64


def bucket(salt: str, unit_id: str) -> float:
    """Map ``(salt, unit_id)`` deterministically into ``[0, 1)``."""
    digest = hashlib.sha256(f"{salt}:{unit_id}".encode()).digest()
    return int.from_bytes(digest[:8], "big") / _UINT64

File: sdks/typescript/src/bucketing.ts
import { createHash } from "node:crypto";

// 2**64 is exactly representable as a double, so dividing by it is an exact exponent
// shift — this is why Number(uint64) / 2**64 matches Python's uint64 / 2**64 bit-for-bit.
const TWO_POW_64 = 2 ** 64;

/**
 * Map (salt, unitId) deterministically into [0, 1).
 * Identical to the Python backend (see ../fixtures/bucketing.json).
 */
export function bucket(salt: string, unitId: string): number {
  const digest = createHash("sha256").update(`${salt}:${unitId}`, "utf8").digest();
  let value = 0n;
  for (let i = 0; i < 8; i++) {
    value = (value << 8n) | BigInt(digest[i]);
  }
  return Number(value) / TWO_POW_64;
}
